lib/pos_mempool.go

```diff
@@ -0,0 +1,446 @@
+	if txn.Hash == nil {
+		return false
+	}
+
+	if _, ok := tr.txnMembership[*txn.Hash]; ok {
+		return false
+	}
+
+	// If the transaction overflows the maximum mempool size, reject it.
+	if tr.totalTxnSize+txn.TxSizeBytes > tr.params.MaxMempoolPosSizeBytes {
+		return false
+	}
+
+	if err := tr.buckets.AddTransaction(txn); err != nil {
+		glog.Errorf("TransactionRegister.AddTransaction: Problem adding txn: %v", err)
+		return false
+	}
+	tr.totalTxnSize += txn.TxSizeBytes
+	tr.txnMembership[*txn.Hash] = struct{}{}
+	return true
+}
+
+// RemoveTransaction removes a transaction from the register.
+func (tr *TransactionRegister) RemoveTransaction(txn *MempoolTx) bool {
+	if _, ok := tr.txnMembership[*txn.Hash]; !ok {
+		return false
+	}
+
+	tr.buckets.RemoveTransaction(txn)
+	delete(tr.txnMembership, *txn.Hash)
+	tr.totalTxnSize -= txn.TxSizeBytes
+	return true
+}
+
+func (tr *TransactionRegister) Empty() bool {
+	return tr.buckets.Empty()
+}
+
+// GetFeeTimeIterator returns an iterator over the transactions in the register. The iterator goes through all transactions
+// as ordered by Fee-Time.
+func (tr *TransactionRegister) GetFeeTimeIterator() *FeeTimeIterator {
+	return tr.buckets.GetFeeTimeIterator()
+}
+
+// GetFeeTimeTransactions returns all transactions in the register ordered by Fee-Time.
+func (tr *TransactionRegister) GetFeeTimeTransactions() []*MempoolTx {
+	txns := []*MempoolTx{}
+	it := tr.GetFeeTimeIterator()
+	for it.Next() {
+		if txn, ok := it.Value(); ok {
+			txns = append(txns, txn)
+		}
+	}
+	return txns
+}
+
+// ========================
+//	FeeBucket
+// ========================
+
+// FeeBucket is a data structure storing time buckets. Time buckets, in turn, store timestamp-ordered mempool transactions
+// that have neighbouring fees. When a transaction is added to the FeeBucket, an appropriate TimeBucket is found or created,
+// depending on the transaction's fee, and the transactions is inserted into it.
+type FeeBucket struct {
+	// timeBucketSet is a set of TimeBuckets. The set is ordered by the TimeBucket's index, based on FeeBucketComparator.
+	timeBucketSet *treeset.Set
+	// timeBucketIndex is a map of TimeBucket indices to TimeBuckets. It is used to quickly find a TimeBucket given its index.
+	timeBucketIndex map[int]*TimeBucket
+
+	globalParams *GlobalParamsEntry
+}
+
+func NewFeeBucket(globalParams *GlobalParamsEntry) *FeeBucket {
+	timeBucketSet := treeset.NewWith(FeeBucketComparator)
+
+	return &FeeBucket{
+		timeBucketSet:   timeBucketSet,
+		timeBucketIndex: make(map[int]*TimeBucket),
+		globalParams:    globalParams,
+	}
+}
+
+// FeeBucketComparator is a comparator function for TimeBuckets. It is used to order TimeBuckets in
+// the FeeBucket's timeBucketSet.
+func FeeBucketComparator(a, b interface{}) int {
+	aVal := a.(*TimeBucket)
+	bVal := b.(*TimeBucket)
+
+	// Determine the TimeBucket with a higher fee range. We can easily find out by comparing TimeBucket indices.
+	if aVal.index < bVal.index {
```

Comment:
Comparing the indices assumes that a and b are part of the same parent FeeBucket instance. Should we compare their fee ranges instead so that this function always works regardless of where a and b were initialized?